src/android_clipboard.rs

```diff
@@ -0,0 +1,93 @@
+use crate::common::{ClipboardProvider, Result};
+use jni::objects::JString;
+use std::ffi::CStr;
+
+pub struct AndroidClipboardContext;
+
+impl ClipboardProvider for AndroidClipboardContext {
+    fn get_contents(&mut self) -> Result<String> {
+        let ctx = ndk_context::android_context();
+        let vm = unsafe { jni::JavaVM::from_raw(ctx.vm().cast()) }?;
+        let env = vm.attach_current_thread()?;
+        let class_ctx = env.find_class("android/content/Context")?;
+        let cb = env.get_static_field(class_ctx, "CLIPBOARD_SERVICE", "Ljava/lang/String;")?;
+        let cb_manager = env
+            .call_method(
+                ctx.context() as jni::sys::jobject,
+                "getSystemService",
+                "(Ljava/lang/String;)Ljava/lang/Object;",
+                &[cb],
+            )?
+            .l()
+            .unwrap();
+
+        let clip_data = env
+            .call_method(cb_manager, "getPrimaryClip", "()Landroid/content/ClipData;", &[])?
+            .l()
+            .unwrap();
```

Comment:
Can this unwrap fail? We should never panic in this method.

Reply:
Yes it won't panic theoretically, I'm 100% sure it will be `JObject`. If not, it must be something wrong with that jvm or the API has changed, this time it should panic. Follows are the same.

Reply:
> Yes it won't panic theoretically

The thing is, it also shouldn't panic practically. We can't afford to crash the entire application of the consumer just because the theory went wrong. And the methods themselves already return a result, so in scenarios where you're not 100% certain that methods will **never** panic, I suggest just returning the error using `?` at basically no cost.